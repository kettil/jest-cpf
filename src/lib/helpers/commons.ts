import { ExtendPromiseType } from '../types';

export const delay = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms, ms));

export const extendPromise = <T>(promise: Promise<T>): ExtendPromiseType<T | Error> => {
  let isPending = true;

  return {
    isPending: () => isPending,
    promise: new Promise<T | Error>((resolve) => {
      promise.then(
        (v: T) => {
          isPending = false;
          resolve(v);
        },
        (error: Error) => {
          isPending = false;
          resolve(error);
        },
      );
    }),
  };
};
