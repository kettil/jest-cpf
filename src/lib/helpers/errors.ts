export class TestError extends Error {
  readonly data: string;

  constructor(message?: string, data = '') {
    super(message);

    this.data = data;
  }
}
