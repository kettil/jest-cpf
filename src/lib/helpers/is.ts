import { CoverageSummaryType } from '../types';

export const isObject = (data: unknown): data is Record<string, unknown> => typeof data === 'object' && data !== null;

export const isSummary = (data: unknown): data is CoverageSummaryType =>
  isObject(data) && typeof data.pct === 'number' && typeof data.total === 'number' && typeof data.covered === 'number';

export const isGlobalThreshold = (threshold: unknown): threshold is { global: Record<string, unknown> } =>
  isObject(threshold) && isObject(threshold.global);
