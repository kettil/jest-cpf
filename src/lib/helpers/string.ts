import { join, dirname, basename } from 'path';

export const cleanFileFromRootPath = (file: string, rootDirectory: string) =>
  file.slice(Math.max(0, rootDirectory.length + 1));

export const cleanFileFromExtensions = (file: string) => {
  const path = dirname(file);
  const base = basename(file);
  const name = base.slice(0, Math.max(0, base.indexOf('.') + 1));

  return join(path, name);
};

export const mapPerLine = (value: string, callback: (line: string) => string) =>
  value
    .split('\n')
    .map(callback)
    .join('\n');

export const padFill = (value: string, length: number, fill: string) =>
  [...new Array(Math.max(0, length - value.length))].fill(fill).join('');

export const padLeft = (value: string, length: number, fill: string): string => padFill(value, length, fill) + value;

export const padRight = (value: string, length: number, fill: string): string => value + padFill(value, length, fill);

export const formatPercent = (value: number): string => {
  const [pre, post = ''] = value.toString().split('.');

  return [padLeft(pre, 3, ' '), padRight(post, 2, '0')].join('.');
};
