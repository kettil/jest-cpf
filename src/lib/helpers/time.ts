import { DateTime } from 'luxon';

export const startTime = () => DateTime.utc();

export const getRunTime = (time: DateTime) => {
  const runtime = DateTime.utc().diff(time, ['minutes', 'seconds']);

  return runtime.toFormat('mm:ss');
};
