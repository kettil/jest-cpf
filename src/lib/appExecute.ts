import cleanUpFiles from './files/cleanUpFiles';
import getFileLists from './files/fileLists';
import { getJestConfig } from './files/jest';
import { cleanFileFromRootPath } from './helpers/string';
import normalizeConfig from './normalize';
import { outputStatus } from './outputs';
import createRunners from './runner/createRunners';
import runTests from './runner/runTests';
import { thresholdDefaultLimits } from './threshold/settings';
import { RunnerType, ArgvObjectType } from './types';

export const getFileListsCallback = (rootDirectory?: string) => {
  if (typeof rootDirectory === 'string') {
    return (path: string) => cleanFileFromRootPath(path, rootDirectory);
  }

  return () => '';
};

const appExecute = async (
  argvArray: string[],
  argvObject: ArgvObjectType,
  cwdDirectory: string,
  cpuCount: number,
  cpuUsed: number,
) => {
  if (typeof argvObject.config !== 'string') {
    throw new TypeError('The Argument "--config" is missing');
  }

  const jestConfig = await getJestConfig(cwdDirectory, argvObject.config);
  const { thresholdLimits, ...config } = normalizeConfig(jestConfig, cwdDirectory);

  const [files, testFiles, runners] = await Promise.all<string[], string[], RunnerType[]>([
    getFileLists([config.rootDir], config.collectCoverageFrom, getFileListsCallback()),
    getFileLists(config.roots, config.testMatch, getFileListsCallback(config.rootDir)),
    createRunners(config, cpuUsed),
  ]);

  const cleanedUpFiles = cleanUpFiles(files, testFiles, config);

  outputStatus(cleanedUpFiles.length, testFiles.length, cpuCount, cpuUsed);

  const thresholdData = await runTests(testFiles, runners, argvArray, argvObject.config, config);

  return { cleanedUpFiles, thresholdData, thresholdLimits: { ...thresholdDefaultLimits, ...thresholdLimits } };
};

export default appExecute;
