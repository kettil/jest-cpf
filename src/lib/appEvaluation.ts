import { DateTime } from 'luxon';
import appExecute from './appExecute';
import { cleanUpByFiles } from './files/cleanUpFiles';
import { getRunTime } from './helpers/time';
import { outputTable, outputMetrics } from './outputs';
import { evaluationFiles } from './threshold/evaluation';
import { thresholdForFilesWithoutTests } from './threshold/settings';
import { UnpackedPromise } from './types';

const appEvaluation = (time: DateTime, data: UnpackedPromise<ReturnType<typeof appExecute>>) => {
  const { cleanedUpFiles, thresholdData, thresholdLimits } = data;

  // remove all files with tests
  cleanUpByFiles(
    cleanedUpFiles,
    thresholdData.map(({ file }) => file),
  ).forEach((file) => {
    // add files without tests
    thresholdData.push({ file, threshold: thresholdForFilesWithoutTests });
  });

  const status = evaluationFiles(thresholdData, thresholdLimits);

  outputTable(thresholdData, thresholdLimits);
  outputMetrics(status, getRunTime(time));

  return status;
};

export default appEvaluation;
