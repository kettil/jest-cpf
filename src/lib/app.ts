import { cpus } from 'os';
import appEvaluation from './appEvaluation';
import appExecute from './appExecute';
import { startTime } from './helpers/time';
import { ArgvObjectType } from './types';

const app = async (argvArray: string[], argvObject: ArgvObjectType) => {
  const time = startTime();

  const cpuCount = cpus().length;
  const cpuUsed = Math.max(1, Math.floor(cpuCount * 0.9));

  const data = await appExecute(argvArray, argvObject, process.cwd(), cpuCount, cpuUsed);
  const status = appEvaluation(time, data);

  // eslint-disable-next-line unicorn/no-process-exit
  process.exit(status.error > 0 ? 1 : 0);
};

export default app;
