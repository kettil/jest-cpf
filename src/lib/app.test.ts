/* eslint-disable @typescript-eslint/unbound-method */
import { cpus } from 'os';
import app from './app';
import appEvaluation from './appEvaluation';
import appExecute from './appExecute';
import { startTime } from './helpers/time';

jest.mock('os');
jest.mock('./helpers/time');
jest.mock('./appExecute');
jest.mock('./appEvaluation');
jest.spyOn(process, 'exit');
jest.spyOn(process, 'cwd');

const cwdMock = process.cwd as jest.Mock;
const cpusMock = cpus as jest.Mock;
const startTimeMock = startTime as jest.Mock;
const appExecuteMock = appExecute as jest.Mock;
const appEvaluationMock = appEvaluation as jest.Mock;

describe('app()', () => {
  beforeEach(() => {
    cwdMock.mockReturnValueOnce('/path/to/project');
    cpusMock.mockReturnValueOnce([1, 2, 3, 4]);
    startTimeMock.mockReturnValueOnce({ time: 'object' });
    appExecuteMock.mockResolvedValueOnce({ data: 'fromExecute' });
  });

  test('it should work with faulty tests', async () => {
    appEvaluationMock.mockReturnValueOnce({ error: 3, success: 97 });

    const argvArray = ['--config', '/path/to/config.json'];
    const argvObject = { config: '/path/to/config.json' };

    const promise = app(argvArray, argvObject);

    await expect(promise).resolves.toBeUndefined();

    expect(cpusMock).toHaveBeenCalledTimes(1);

    expect(appExecuteMock).toHaveBeenCalledTimes(1);
    expect(appExecuteMock).toMatchSnapshot('appExecute');

    expect(appEvaluationMock).toHaveBeenCalledTimes(1);
    expect(appEvaluationMock).toMatchSnapshot('appEvaluation');

    expect(process.exit).toHaveBeenCalledTimes(1);
    expect(process.exit).toHaveBeenCalledWith(1);
  });

  test('it should work without faulty tests', async () => {
    appEvaluationMock.mockReturnValueOnce({ error: 0, success: 100 });

    const argvArray = ['--config', '/path/to/config.json'];
    const argvObject = { config: '/path/to/config.json' };

    const promise = app(argvArray, argvObject);

    await expect(promise).resolves.toBeUndefined();

    expect(process.exit).toHaveBeenCalledTimes(1);
    expect(process.exit).toHaveBeenCalledWith(0);
  });
});
