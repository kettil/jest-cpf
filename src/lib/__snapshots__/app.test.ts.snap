// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app() it should work with faulty tests: appEvaluation 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "time": "object",
      },
      Object {
        "data": "fromExecute",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "error": 3,
        "success": 97,
      },
    },
  ],
}
`;

exports[`app() it should work with faulty tests: appExecute 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        "--config",
        "/path/to/config.json",
      ],
      Object {
        "config": "/path/to/config.json",
      },
      "/path/to/project",
      4,
      3,
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;
